File: src/main/java/com/temelyan/pomoapp/util/PasswordUtil.java
package com.temelyan.pomoapp.util;

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.util.StringUtils;

import java.util.regex.Pattern;

public class PasswordUtil {
    private static final PasswordEncoder PASSWORD_ENCODER = new BCryptPasswordEncoder();
    private static final Pattern BCRYPT_PATTERN = Pattern.compile("\\A\\$2a?\\$\\d\\d\\$[./0-9A-Za-z]{53}");

    public static PasswordEncoder getPasswordEncoder() {
        return PASSWORD_ENCODER;
    }

    public static String encode(String newPassword) {
        if (!StringUtils.hasText(newPassword)) {
            return null;
        }
        if (isEncoded(newPassword)) {
            return newPassword;
        }
        return PASSWORD_ENCODER.encode(newPassword);
    }

    private static boolean isEncoded(String newPassword) {
        return BCRYPT_PATTERN.matcher(newPassword).matches();
    }
}

File: src/main/java/com/temelyan/pomoapp/util/PasswordUtil.kt
package com.temelyan.pomoapp.util

import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.util.StringUtils

import java.util.regex.Pattern

object PasswordUtil {
    private val BCRYPT_PATTERN = Pattern.compile("\\A\\$2a?\\$\\d\\d\\$[./0-9A-Za-z]{53}")

    fun encode(newPassword: String, passwordEncoder: PasswordEncoder): String? {
        if (!StringUtils.hasText(newPassword)) {
            return null
        }
        return if (isEncoded(newPassword)) {
            newPassword
        } else passwordEncoder.encode(newPassword)
    }

    private fun isEncoded(newPassword: String): Boolean {
        return BCRYPT_PATTERN.matcher(newPassword).matches()
    }
}
